File: ardalink-api/src/routes/satellite.ts
/**
 * Satellite routes — live GEE VCI and MODIS NDVI data.
 *
 * Proxies requests to the ArdaLink engine's `/api/v1/satellite/*` endpoints,
 * which wrap the Google Earth Engine pipeline.
 *
 * Routes:
 * - GET /api/satellite/vci?ward=<ward_id>
 * - POST /api/satellite/trigger
 */

import { Router, type IRouter, type Request, type Response } from "express";
import { fetchSatelliteVCI, triggerSatelliteRefresh, type VCISnapshot } from "../lib/engine.js";
import { logger } from "../lib/logger.js";
import { withTenantContext } from "../lib/tenancy-context.js";
import { satelliteSnapshotsTable } from "@workspace/db";

const router: IRouter = Router();

/**
 * GET /api/satellite/vci?ward=<ward_id>
 *
 * Returns the live Vegetation Condition Index for a single ward.
 * Proxies to the engine and returns 503 when GEE is not configured.
 */
router.get("/satellite/vci", async (req: Request, res: Response): Promise<void> => {
  const wardId = req.query.ward as string;
  if (!wardId) {
    res.status(400).json({ error: "missing_ward", message: "ward query parameter is required" });
    return;
  }

  try {
    // The engine's tenancy middleware requires an attested X-Tenant-ID header
    // when TENANT_ATTESTATION_SECRET is set. Fall back to "isiolo" when the
    // caller has no bound tenant (public/demo requests).
    const tenantId = req.tenant?.tenant_id ?? "isiolo";
    const vci = await fetchSatelliteVCI(wardId, tenantId);

    if (!vci) {
      res.status(503).json({
        error: "engine_unreachable",
        message: "ArdaLink engine is not responding. Ensure it's running on the configured port.",
      });
      return;
    }

    // Also write to satellite_snapshots table for persistence
    await withTenantContext(tenantId, async (tx) => {
      await tx
        .insert(satelliteSnapshotsTable)
        .values({
          tenantId,
          newestImageDate: vci.captured_at.split("T")[0] ?? new Date().toISOString().split("T")[0],
          result: {
            ward_id: vci.ward_id,
            ward_name: vci.ward_name,
            vci: vci.vci,
            ndvi_now: vci.ndvi_now,
            ndvi_min: vci.ndvi_min,
            ndvi_max: vci.ndvi_max,
            urban_masked: vci.urban_masked,
            prosopis_factor: vci.prosopis_factor,
            captured_at: vci.captured_at,
          },
        })
        .onConflictDoNothing(); // Don't overwrite if recent snapshot exists
    });

    logger.info({ wardId, vci: vci.vci }, "[Satellite] VCI fetched and cached");
    res.json(vci);
  } catch (err: unknown) {
    logger.error({ err, wardId }, "[Satellite] VCI fetch failed");
    res.status(500).json({ error: "satellite_fetch_failed", message: String(err) });
  }
});

/**
 * POST /api/satellite/trigger?dryRun=false
 *
 * Triggers VCI fetch for all demo wards (Bula Pesa, Garbatulla, Kinna).
 * Used by the satellite scheduler and manual trigger from the dashboard.
 */
router.post("/satellite/trigger", async (req: Request, res: Response): Promise<void> => {
  const dryRun = req.query.dryRun === "true" || req.body?.dryRun === true;

  try {
    const tenantId = req.tenant?.tenant_id ?? "isiolo";
    const result = await triggerSatelliteRefresh(dryRun, tenantId);

    if (!result) {
      res.status(503).json({
        error: "engine_unreachable",
        message: "ArdaLink engine is not responding.",
      });
      return;
    }

    // If successful and not a dry run, write results to satellite_snapshots
    if (result.status === "success" && result.results && !dryRun) {
      for (const [wardId, snapshot] of Object.entries(result.results)) {
        if (!snapshot) continue;

        await withTenantContext(tenantId, async (tx) => {
          await tx
            .insert(satelliteSnapshotsTable)
            .values({
              tenantId,
              newestImageDate: new Date().toISOString().split("T")[0],
              result: snapshot,
            })
            .onConflictDoNothing();
        });
      }

      logger.info(
        { wards: result.wards, status: result.status },
        "[Satellite] Trigger completed and snapshots written",
      );
    }

    res.json(result);
  } catch (err: unknown) {
    logger.error({ err }, "[Satellite] Trigger failed");
    res.status(500).json({ error: "satellite_trigger_failed", message: String(err) });
  }
});

/**
 * GET /api/satellite/snapshots?limit=10
 *
 * Returns recent satellite snapshots from the database.
 * Useful for the dashboard to display historical VCI trends.
 */
router.get("/satellite/snapshots", async (req: Request, res: Response): Promise<void> => {
  const limit = Math.min(100, Math.max(1, Number(req.query.limit ?? 10)));

  try {
    const tenantId = req.tenant?.tenant_id;
    if (!tenantId) {
      res.status(401).json({ error: "unauthorized", message: "Tenant context required" });
      return;
    }

    const snapshots = await withTenantContext(tenantId, async (tx) => {
      return tx
        .select({
          id: satelliteSnapshotsTable.id,
          capturedAt: satelliteSnapshotsTable.capturedAt,
          newestImageDate: satelliteSnapshotsTable.newestImageDate,
          result: satelliteSnapshotsTable.result,
        })
        .from(satelliteSnapshotsTable)
        .orderBy(satelliteSnapshotsTable.capturedAt)
        .limit(limit);
    });

    res.json({
      tenant_id: tenantId,
      count: snapshots.length,
      snapshots,
    });
  } catch (err: unknown) {
    logger.error({ err }, "[Satellite] Snapshots fetch failed");
    res.status(500).json({ error: "snapshots_fetch_failed", message: String(err) });
  }
});

export default router;

File: ardalink-engine/ardalink_engine/src/api/satellite.py
"""Satellite endpoints — live GEE VCI and MODIS NDVI data.

Wraps the existing GEE pipeline in :mod:`src.pipeline.satellite` and exposes
HTTP endpoints for the API to consume.

Two endpoints:

* ``GET /api/v1/satellite/vci``  — live Vegetation Condition Index for one ward
* ``POST /api/v1/satellite/trigger`` — fetch VCI for all demo wards

Both endpoints call into ``src.pipeline.satellite.fetch_vegetation_index``,
which handles GEE initialization, MODIS NDVI aggregation, urban masking,
and the Prosopis juliflora penalty.
"""

from __future__ import annotations

from datetime import UTC, datetime

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from ..geo.wards import WARDS
from ..logging_config import get_logger
from ..pipeline.gee import GEEInitError, GEENotConfigured

logger = get_logger("ardalink.api.satellite")

router = APIRouter(prefix="/api/v1/satellite", tags=["satellite"])

# The 5 active Isiolo Sub-County wards (canonical since 2026-07 tenant
# retirement of Garbatulla + Merti). Slugs map to WARDS display-name keys
# via `_SLUG_TO_NAME` below — the resolver was previously falling back
# to title-case conversion, which produced "Bula Pesa" instead of the
# canonical "Bulla Pesa" (double-L) and 404'd every trigger.
DEMO_WARDS = ["bulla-pesa", "wabera", "ngare-mara", "burat", "oldonyiro"]

# Canonical slug → WARDS display-name lookup. Restricted to the 5 active
# Isiolo Sub-County wards. Retired demo tenants (garbatulla, merti) and
# non-active neighbouring wards (kinna, chari, cherab, sericho) were
# dropped 2026-07-21 — any caller still naming them will 404, which is
# preferable to silently returning data for a ward not in active_wards.
_SLUG_TO_NAME: dict[str, str] = {
    "bulla-pesa": "Bulla Pesa",
    "bula-pesa": "Bulla Pesa",  # legacy single-L spelling
    "wabera": "Wabera",
    "ngare-mara": "Ngare Mara",
    "burat": "Burat",
    "oldonyiro": "Oldonyiro",
    "oldo-nyiro": "Oldonyiro",
}


class VCISnapshot(BaseModel):
    """Live Vegetation Condition Index for a single ward."""

    ward_id: str
    ward_name: str
    vci: float
    ndvi_now: float | None
    ndvi_min: float | None
    ndvi_max: float | None
    urban_masked: bool
    prosopis_factor: float
    image_dates: list[str] | None
    captured_at: str


class TriggerResponse(BaseModel):
    """Response from POST /api/v1/satellite/trigger."""

    status: str
    wards: list[str]
    started_at: str
    results: dict[str, dict] | None
    error: str | None


def _resolve_ward(ward_id: str) -> str:
    """Resolve a ward ID to a normalized ward name.

    Accepts either a slug (e.g. 'bula-pesa') or a display name
    (e.g. 'Bulla Pesa'). Returns the canonical name used in WARDS keys.
    """
    # Direct match
    if ward_id in WARDS:
        return ward_id

    # Curated slug alias table — handles the double-L "Bulla" vs
    # slug "bula-pesa" mismatch and any legacy renames.
    slug = ward_id.lower().strip()
    if slug in _SLUG_TO_NAME:
        name = _SLUG_TO_NAME[slug]
        if name in WARDS:
            return name

    # Case-insensitive display-name match
    for key in WARDS:
        if key.lower() == slug:
            return key

    # Fallback: hyphen→space title-case (only correct for wards whose
    # slug and display name are trivially related — kept for safety net).
    if "-" in ward_id:
        converted = ward_id.replace("-", " ").title()
        if converted in WARDS:
            return converted
        for key in WARDS:
            if key.lower() == converted.lower():
                return key

    raise HTTPException(
        status_code=404,
        detail={"error": "unknown_ward", "ward_id": ward_id, "known_wards": list(WARDS.keys())},
    )


@router.get("/vci", response_model=VCISnapshot)
def get_vci(
    ward_id: str = Query(..., description="Ward slug (e.g. 'bula-pesa') or display name"),
) -> VCISnapshot:
    """Return the live Vegetation Condition Index for a single ward.

    Calls ``src.pipeline.satellite.fetch_vegetation_index``, which:
    * Initializes GEE (lazy, thread-safe)
    * Fetches MODIS MOD13Q1 NDVI for the ward
    * Computes VCI = 100 * (NDVI_now - NDVI_min) / (NDVI_max - NDVI_min)
    * Applies urban mask (ESA WorldCover built-up pixels)
    * Applies Prosopis juliflora penalty (0.85 factor)

    Returns HTTP 503 when GEE is not configured.
    """
    try:
        resolved_ward = _resolve_ward(ward_id)
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(status_code=400, detail={"error": "invalid_ward", "message": str(exc)}) from exc

    try:
        from ..pipeline.satellite import fetch_vegetation_index

        result = fetch_vegetation_index(resolved_ward)
        ward = WARDS[resolved_ward]

        return VCISnapshot(
            ward_id=resolved_ward,
            ward_name=ward.display_name if hasattr(ward, "display_name") else resolved_ward,
            vci=result["vci"],
            ndvi_now=result["ndvi_now"],
            ndvi_min=result["ndvi_min"],
            ndvi_max=result["ndvi_max"],
            urban_masked=result["urban_masked"],
            prosopis_factor=result["prosopis_factor"],
            image_dates=None,  # GEE doesn't return image dates in the current implementation
            captured_at=datetime.now(UTC).isoformat(),
        )
    except GEENotConfigured as exc:
        logger.error("GEE not configured: %s", exc)
        raise HTTPException(
            status_code=503,
            detail={
                "error": "gee_not_configured",
                "message": "Google Earth Engine credentials are not configured. Set GEE_PRIVATE_KEY and GEE_SERVICE_ACCOUNT.",
            },
        ) from exc
    except GEEInitError as exc:
        logger.error("GEE initialization failed: %s", exc)
        raise HTTPException(
            status_code=503,
            detail={
                "error": "gee_init_failed",
                "message": f"Earth Engine initialization failed: {exc}",
            },
        ) from exc
    except Exception as exc:
        logger.exception("Unexpected error fetching VCI for ward '%s'", ward_id)
        raise HTTPException(
            status_code=500,
            detail={"error": "internal_error", "message": f"Failed to fetch VCI: {exc}"},
        ) from exc


@router.post("/trigger", response_model=TriggerResponse)
def trigger_all_wards(
    dry_run: bool = Query(False, description="If true, skip actual GEE calls"),
) -> TriggerResponse:
    """Fetch VCI for all demo wards (Bula Pesa, Garbatulla, Kinna).

    Calls ``fetch_vegetation_index`` for each ward in parallel and returns
    the combined results. Use this for scheduled refreshes or manual
    triggering from the operator dashboard.

    Set ``dry_run=true`` to validate the endpoint without making GEE calls.
    """
    started_at = datetime.now(UTC).isoformat()

    if dry_run:
        return TriggerResponse(
            status="dry_run",
            wards=DEMO_WARDS,
            started_at=started_at,
            results=None,
            error=None,
        )

    try:
        from ..pipeline.satellite import fetch_vegetation_index

        results: dict[str, dict] = {}
        errors: dict[str, str] = {}

        for ward_slug in DEMO_WARDS:
            try:
                # Resolve slug to canonical ward name before calling fetch_vegetation_index
                ward_name = _resolve_ward(ward_slug)
                result = fetch_vegetation_index(ward_name)
                results[ward_slug] = result
                logger.info("VCI fetch succeeded for %s: %.1f", ward_slug, result["vci"])
            except HTTPException:
                # Re-raise HTTPExceptions (including unknown_ward) as-is
                raise
            except GEENotConfigured:
                errors[ward_slug] = "GEE not configured"
                raise  # Re-raise so caller sees 503
            except GEEInitError as exc:
                errors[ward_slug] = f"GEE init failed: {exc}"
                raise  # Re-raise so caller sees 503
            except Exception as exc:
                errors[ward_slug] = str(exc)
                logger.warning("VCI fetch failed for %s: %s", ward_slug, exc)

        if not results:
            return TriggerResponse(
                status="failed",
                wards=DEMO_WARDS,
                started_at=started_at,
                results=None,
                error="All wards failed: " + ", ".join(f"{k}={v}" for k, v in errors.items()),
            )

        return TriggerResponse(
            status="success",
            wards=list(results.keys()),
            started_at=started_at,
            results=results,
            error=None if len(results) == len(DEMO_WARDS) else f"Partial failures: {errors}",
        )

    except GEENotConfigured as exc:
        raise HTTPException(
            status_code=503,
            detail={
                "error": "gee_not_configured",
                "message": "Google Earth Engine credentials are not configured.",
            },
        ) from exc
    except GEEInitError as exc:
        raise HTTPException(
            status_code=503,
            detail={
                "error": "gee_init_failed",
                "message": f"Earth Engine initialization failed: {exc}",
            },
        ) from exc
    except Exception as exc:
        logger.exception("Unexpected error in trigger_all_wards")
        raise HTTPException(
            status_code=500,
            detail={"error": "internal_error", "message": f"Trigger failed: {exc}"},
        ) from exc
